tests/api/test_remote_engine.py

```diff
@@ -207,3 +208,135 @@ def test_default_compiler(self, prog, monkeypatch, caplog):
 
         assert engine.device_spec.default_compiler == "Xunitary"
         assert caplog.records[-1].message == "Compiling program for device X8_01 using compiler Xunitary."
+
+    class MockProgram:
+        """A mock program for testing"""
+        def __init__(self):
+            self.run_options = {}
+
+    def test_compile_device_invalid_device_error(self, prog, monkeypatch, caplog):
+        """Tests that an error is raised if the program was compiled for
+        another device and recompilation was not requested."""
+        caplog.set_level(logging.INFO)
+        test_device_dict = mock_device_dict.copy()
+        test_device_dict["compiler"] = []
+
+        monkeypatch.setattr(Connection, "create_job", lambda self, target, program, run_options: program)
+        monkeypatch.setattr(Connection, "_get_device_dict", lambda *args: test_device_dict)
+        monkeypatch.setattr(Program, "compile", lambda *args, **kwargs: self.MockProgram())
+
+        # Setting compile_info with a dummy devicespec and compiler name
+        X8_spec = DeviceSpec(target="DummyDevice", connection=None, spec=None)
+        prog._compile_info = (X8_spec, "dummy_compiler")
+
+        engine = sf.RemoteEngine("X8")
+        with pytest.raises(
+            ValueError, match="Cannot use program compiled"
+        ):
+            program = engine.run_async(prog, shots=10)
+
+    def test_compile(self, prog, monkeypatch, caplog):
+        """Tests that compilation happens by default if no compile_info was
+        specified when call run_async."""
+        caplog.set_level(logging.INFO)
+        test_device_dict = mock_device_dict.copy()
+        test_device_dict["compiler"] = []
+
+        monkeypatch.setattr(Connection, "create_job", lambda self, target, program, run_options: program)
+        monkeypatch.setattr(Connection, "_get_device_dict", lambda *args: test_device_dict)
+        monkeypatch.setattr(Program, "compile", lambda *args, **kwargs: self.MockProgram())
+
+        # Leaving compile_info as None
+        assert prog.compile_info == None
+
+        engine = RemoteEngine("X8")
+        program = engine.run_async(prog, shots=10)
+
+        assert isinstance(program, self.MockProgram)
+        assert caplog.records[-1].message == "Compiling program for device X8_01 using compiler Xunitary."
+
+    def test_recompilation(self, prog, monkeypatch, caplog):
+        """Test that recompilation happens when the recompile keyword argument
+        was set to True."""
+        compiler = "Xunitary"
+
+        caplog.set_level(logging.INFO)
+        test_device_dict = mock_device_dict.copy()
+        test_device_dict["compiler"] = compiler
+
+        monkeypatch.setattr(Connection, "create_job", lambda self, target, program, run_options: program)
+        monkeypatch.setattr(Connection, "_get_device_dict", lambda *args: test_device_dict)
+
+        # This will be popped
```

Comment:
Is this comment a remnant from when `recompile` was part of `compile_options`? :thinking:

Reply:
Yes, thanks for catching it!